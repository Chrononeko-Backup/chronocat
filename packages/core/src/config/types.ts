export interface ChronocatConfig {
  /**
   * @title 启用
   *
   * @description 是否启用这个功能。
   *
   * @default true
   */
  enable?: boolean

  /**
   * @title 服务
   *
   * @description Chronocat 需要启动的服务。
   *
   * @default []
   */
  servers?: (
    | ChronocatRedServerConfig
    | ChronocatSatoriServerConfig
    | ChronocatSatoriWebHookConfig
  )[]

  /**
   * @title 账号配置
   *
   * @description 账号专属的配置。
   *
   * @default {}
   */
  overrides?: {
    [account: string]: ChronocatCurrentConfig
  }
}

export interface ChronocatCurrentConfig {
  /**
   * @title 启用
   *
   * @description 是否启用这个功能。
   */
  enable?: boolean

  /**
   * @title 服务
   *
   * @description Chronocat 需要启动的服务。
   *
   * @default []
   */
  servers?: (
    | ChronocatRedServerConfig
    | ChronocatSatoriServerConfig
    | ChronocatSatoriWebHookConfig
  )[]
}

export interface ChronocatServerConfig {
  /**
   * @title 启用
   *
   * @description 是否启用这个功能。
   *
   * @default true
   */
  enable?: boolean

  /**
   * @title 密码
   *
   * @description 服务的密码。客户端使用服务时需要提供！
   */
  token: string
}

export interface ChronocatHttpServerConfig {
  /**
   * @title 监听地址
   *
   * @description 服务器监听的地址。
   *
   * 如果你不知道这是什么，那么不填此项即可！
   *
   * @default "0.0.0.0"
   */
  listen?: string
}

export interface ChronocatRedServerConfig
  extends ChronocatServerConfig,
    ChronocatHttpServerConfig {
  type: 'red'

  /**
   * @title 监听端口
   *
   * @description 服务器监听的端口。
   *
   * 默认为 16530。
   *
   * @TJS-type integer
   * @default 16530
   * @minimum 1
   * @maximum 65535
   */
  port?: number
}

export interface ChronocatSatoriEventsConfig {
  /**
   * @title 客户端访问 Satori 服务时的地址
   *
   * @description 客户端将会使用这个地址访问 Satori。
   *
   * 大多数情况下可以不填写此配置，但如果你的 Satori
   * 服务与客户端框架位于不同的网络主机（如 Docker
   * 环境），那么你需要填写客户端访问 Satori 时的地址。
   *
   * 举两个例子：
   *
   * - 当 Satori 服务与客户端分别位于局域网下的两台主机时，self_url
   * 应当形如：
   *
   * http://192.168.1.2:5500
   *
   * - 当网络由 Docker Compose 或 Kubernetes 管理时，self_url
   * 应当形如：
   *
   * http://chronocat.local:5500
   *
   * 判断地址是否正确的方法：
   *
   * 从客户端所在的网络环境直接访问此地址，应当看到
   * Satori 返回的「恭喜」的 HTML 页面。
   *
   * @TJS-pattern ^https?:\/\/[A-Za-z0-9:.]+$
   * @default "https://chronocat.vercel.app"
   */
  self_url?: string

  /**
   * @title Satori 服务声明的平台
   *
   * @description Satori 对外宣称的自己所处的平台。
   *
   * 大多数情况下均无需填写此配置。
   *
   * @default "chronocat"
   */
  platform?: string
}

export interface ChronocatSatoriServerConfig
  extends ChronocatServerConfig,
    ChronocatHttpServerConfig,
    ChronocatSatoriEventsConfig {
  type: 'satori'

  /**
   * @title 监听端口
   *
   * @description 服务器监听的端口。
   *
   * 默认为 5500。
   *
   * @TJS-type integer
   * @default 5500
   * @minimum 1
   * @maximum 65535
   */
  port?: number
}

export interface ChronocatSatoriWebHookConfig
  extends ChronocatSatoriEventsConfig {
  type: 'satori_webhook'

  /**
   * @title WebHook URL
   *
   * @description WebHook 推送事件的目标 URL。
   */
  url: string

  /**
   * @title 启用
   *
   * @description 是否启用这个功能。
   *
   * @default true
   */
  enable?: boolean

  /**
   * @title WebHook 推送鉴权密码
   *
   * @description
   *
   * WebHook 推送事件时，会设置 Authorization 头为
   * Bearer 加上此鉴权密码。
   *
   * 强烈建议为 WebHook 服务设置单独的鉴权密码。
   *
   * @default ""
   */
  token?: string
}
