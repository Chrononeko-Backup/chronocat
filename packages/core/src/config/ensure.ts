// https://github.com/import-js/eslint-plugin-import/issues/2802
// eslint-disable-next-line import/no-unresolved
import { app } from 'electron'
import { mkdir, readFile, stat, writeFile } from 'node:fs/promises'
import { join } from 'node:path'
import defaultConfig from '../../../docs/static/chronocat.yml'
import { baseDir, legacyBaseDir } from '../utils/baseDir'
import { generateToken } from '../utils/token'

export const ensureConfig = async () => {
  const configDir = join(baseDir, 'config')
  const configPath = join(configDir, 'chronocat.yml')
  const legacyTokenPath = join(legacyBaseDir, 'RED_PROTOCOL_TOKEN')

  if (!(await exists(configPath))) {
    // 配置文件不存在，生成新配置
    let newToken =
      app.commandLine.getSwitchValue('chrono-default-token') ||
      process.env['CHRONO_DEFAULT_TOKEN'] ||
      generateToken()

    // 检查旧 token 文件
    if (await exists(legacyTokenPath))
      // 旧 token 文件存在，保持 token 不变
      newToken = (await readFile(legacyTokenPath, 'utf-8')).trim()
    else {
      // 旧 token 文件不存在
      // 为兼容性考虑，将生成的 token 写入旧 token 文件
      // TODO: 0.0.46 版本稳定后删除整个 else 块
      await mkdir(legacyBaseDir, {
        recursive: true,
      })
      await writeFile(legacyTokenPath, newToken)
    }

    const newConfig = defaultConfig.replaceAll(
      'DEFINE_CHRONO_TOKEN',
      `'${newToken}'`,
    )
    await mkdir(configDir, {
      recursive: true,
    })
    await writeFile(configPath, newConfig)
  }
}

async function exists(path: string): Promise<boolean> {
  try {
    await stat(path)
  } catch (_) {
    return false
  }
  return true
}
