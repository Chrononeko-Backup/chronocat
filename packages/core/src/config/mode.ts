// eslint-disable-next-line import/no-unresolved
import { app } from 'electron'

let modes: string[] | undefined = undefined

const splitModes = (modes: string) => modes.split(',').filter(Boolean)

const loadModes = () => {
  const result = splitModes(
    app.commandLine.getSwitchValue('chrono-mode'),
  ).concat(splitModes(process.env['CHRONO_MODE'] || ''))

  if (result.includes('headless3')) result.push('login')
  if (result.includes('headless4')) result.push('login')

  return [...new Set(result)]
}

export const getChronocatModes = () => {
  if (!modes) modes = loadModes()
  return modes
}

export const isChronocatMode = (mode: string) =>
  getChronocatModes().includes(mode)
