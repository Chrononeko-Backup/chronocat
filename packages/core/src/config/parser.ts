import type { ValidateFunction } from 'ajv'
import localize from 'ajv-i18n/localize/zh'
import type { ChronocatConfig } from './types'
import { validate } from './validate'

export const parseConfig = (config: unknown, uin: string) => {
  if (!(validate as ValidateFunction<ChronocatConfig>)(config)) {
    localize((validate as ValidateFunction<ChronocatConfig>).errors)

    let e = 'Chronocat 加载配置失败，请检查 chronocat.yml。详细信息：\n'
    let i = 0
    for (const error of (validate as ValidateFunction<ChronocatConfig>).errors!)
      e += `问题 ${++i}：${error.schemaPath}：${error.message}`

    throw e
  }

  const result = Object.assign({}, config, config.overrides?.[uin])
  if ('overrides' in result) delete result.overrides

  return result
}
