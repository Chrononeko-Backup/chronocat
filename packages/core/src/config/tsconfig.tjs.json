{
  "compilerOptions": {
    "target": "es5",
  },
  "include": [
    "types.ts",
  ]
}
