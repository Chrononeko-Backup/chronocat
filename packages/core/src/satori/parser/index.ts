import Element from '@satorijs/element'
import { Buffer } from 'node:buffer'
import type { ChronocatSatoriEventsConfig } from '../../config/types'
import type { Message as RedMessage } from '../../red'
import { ChatType, MsgType, SendType } from '../../red'
import type { Channel, Event, Guild, GuildMember } from '../types'
import { ChannelType } from '../types'
import { parseMsgTypes } from './msgt'

export const buildParser =
  (selfId: string, config: ChronocatSatoriEventsConfig) =>
  (message: RedMessage) =>
    parseMessageRecv(selfId, config, message)

export const parseMessageRecv = async (
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  message: RedMessage,
) => {
  const parsed = await parseMessage(self_id, config, message)

  if (!parsed) return undefined

  const result: Event[] = []

  for (const event of parsed) {
    if (!event.message?.id && !event.user?.id) {
      result.push({
        id: undefined as unknown as number,
        platform: config.platform!,
        self_id,
        timestamp: Number(message.msgTime) * 1000,

        type: 'chrono-unsafe-warning-2127',
      })
      continue
    } else if (!event.message?.id)
      result.push({
        id: undefined as unknown as number,
        platform: config.platform!,
        self_id,
        timestamp: Number(message.msgTime) * 1000,

        type: 'chrono-unsafe-warning-2128',
        user: {
          id: event.user?.id as string,
          name: event.user?.name as string,
        },
      })
    else if (!event.user?.id)
      result.push({
        id: undefined as unknown as number,
        platform: config.platform!,
        self_id,
        timestamp: Number(message.msgTime) * 1000,

        type: 'chrono-unsafe-warning-2129',
        user: {
          id: undefined as unknown as string,
          name: event.user?.name as string,
        },
        message: {
          id: event.message?.id,
          content: undefined as unknown as string,
        },
      })
    else if (!event.user?.name && !event.member?.name)
      result.push({
        id: undefined as unknown as number,
        platform: config.platform!,
        self_id,
        timestamp: Number(message.msgTime) * 1000,

        type: 'chrono-unsafe-warning-2130',
        user: {
          id: event.user?.id,
          name: undefined as unknown as string,
        },
        message: {
          id: event.message?.id,
          content: undefined as unknown as string,
        },
      })

    result.push(event)
  }

  return result
}

export const parseMessage = async (
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  message: RedMessage,
) => {
  const event: Event = {
    id: undefined as unknown as number,
    type: undefined as unknown as string,

    platform: config.platform!,
    self_id,
    timestamp: Number(message.msgTime) * 1000,
  }

  event.user = {
    id: message.senderUin,
    name: (message.sendNickName || undefined) as unknown as string,
    avatar: `http://q.qlogo.cn/headimg_dl?dst_uin=${message.senderUin}&spec=640`,
  }

  const ntMsgTypes = parseMsgTypes(message)

  // 无论哪种消息都有 Channel 和 User
  event.channel = {} as Channel

  // 判断消息来源
  switch (ntMsgTypes.chatType) {
    case ChatType.Private:
      event.channel.type = ChannelType.DIRECT
      event.channel.id = `private:${event.user.id}`
      event.channel.name = event.user.name
      break

    case ChatType.Group:
      // Guild 和 Member 只有群聊有
      event.guild = {} as Guild
      event.member = {} as GuildMember

      if (message.sendMemberName) event.member.name = message.sendMemberName

      event.channel.type = ChannelType.TEXT
      event.channel.id = event.guild.id = message.peerUid
      event.channel.name = event.guild.name = message.peerName
      event.guild.avatar = `https://p.qlogo.cn/gh/${message.peerUid}/${message.peerUid}/640`
      break
  }

  if (
    ntMsgTypes.msgType === MsgType.Normal &&
    ntMsgTypes.subMsgType.face &&
    message.elements.some((x) => x?.faceElement?.faceType === 5)
  )
    // 窗口抖动 / 手机按钮戳一戳
    // TODO
    return undefined
  else if (
    ntMsgTypes.msgType === MsgType.Ark &&
    message.subMsgType === 0 &&
    ntMsgTypes.sendType === SendType.Normal
  )
    // ARK 卡片消息，elementType = 10
    // 不处理
    return undefined
  else if (
    ntMsgTypes.msgType === MsgType.Normal ||
    ntMsgTypes.msgType === MsgType.Value3 ||
    ntMsgTypes.msgType === MsgType.Ptt ||
    ntMsgTypes.msgType === MsgType.Video ||
    ntMsgTypes.msgType === MsgType.WithRecords ||
    ntMsgTypes.msgType === MsgType.Vaule17
  )
    return parseChatMessage(self_id, config, event, message).then((x) => [
      x[0],
      ...x[1],
    ])
  // else if (event.__CHRONO_UNSAFE_NTMSGTYPES__.subMsgType.multiForward)
  //   // 合并转发消息
  //   // multiForwardMsgElement（elementType = 16）内并不带有合并转发的全部内容，
  //   // 需要后续通过 API 再请求
  //   // 考虑到合并转发消息解析需求较少，不调度此 session
  //   return undefined
  else if (
    ntMsgTypes.msgType === MsgType.System && // 5
    message.subMsgType === 8 && // 8
    ntMsgTypes.sendType === SendType.System && // 3
    message.elements[0]!.elementType === 8 && // 8
    message.elements[0]!.grayTipElement!.subElementType === 4 && // 4
    message.elements[0]!.grayTipElement!.groupElement!.type === 1 // 1
  )
    // 新人自行入群
    return await parseGuildMemberAddedMessage(self_id, config, event, message)
  else if (
    ntMsgTypes.msgType === MsgType.System && // 5
    message.subMsgType === 8 && // 8
    ntMsgTypes.sendType === SendType.System && // 3
    message.elements[0]!.elementType === 8 && // 8
    message.elements[0]!.grayTipElement!.subElementType === 4 && // 4
    message.elements[0]!.grayTipElement!.groupElement!.type === 8 // 8
  )
    // 他人被禁言
    return await parseGuildMemberMuteMessage(self_id, config, event, message)
  else if (
    ntMsgTypes.msgType === MsgType.System && // 5
    message.subMsgType === 8 && // 8
    ntMsgTypes.sendType === SendType.System && // 3
    message.elements[0]!.elementType === 8 && // 8
    message.elements[0]!.grayTipElement!.subElementType === 4 && // 4
    message.elements[0]!.grayTipElement!.groupElement!.type === 5 // 5
  )
    // 群名称变更
    return undefined
  else if (
    ntMsgTypes.msgType === MsgType.System && // 5
    message.subMsgType === 12 && // 12
    ntMsgTypes.sendType === SendType.System && // 3
    message.elements[0]!.elementType === 8 && // 8
    message.elements[0]!.grayTipElement!.subElementType === 12 && // 12
    message.elements[0]!.grayTipElement!.xmlElement!.busiType === '1' && // 1
    message.elements[0]!.grayTipElement!.xmlElement!.busiId === '10145' // 10145
  )
    // 旧版群成员邀请新人入群
    return await parseGuildMemberAddedLegacyInviteMessage(
      self_id,
      config,
      event,
      message,
    )
  else if (
    ntMsgTypes.msgType === MsgType.System &&
    message.subMsgType === 17 &&
    ntMsgTypes.sendType === SendType.System
  )
    // 群主禁止群内临时通话
    // 群主禁止群内发起新的群聊
    return undefined

  return undefined
}

/**
 * 解析聊天消息。
 *
 * @remarks
 * 在消息没有除了消息元素以外的其他属性需要处理的情况下，直接使用此方法。
 */
async function parseChatMessage(
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  event: Event,
  message: RedMessage,
) {
  const [elements, extraEvents] = await parseElements(self_id, config, message)
  event.type = 'message-created'
  event.message = {
    id: message.msgId,
    content: elements.join(''),
  }
  return [event, extraEvents] as const
}

/**
 * 解析新人自行入群消息。
 *
 * @remarks
 * 通过解析 `grayTipElement`（elementType = 8）中的
 * `groupElement`（subElementType = 4）即可直接提取 QQ 号。
 */
async function parseGuildMemberAddedMessage(
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  event: Event,
  message: RedMessage,
) {
  const [event2, extraEvents] = await parseChatMessage(
    self_id,
    config,
    event,
    message,
  )
  event2.type = 'guild-member-added'

  event2.operator = {
    id: message.elements[0]!.grayTipElement!.groupElement!.adminUin!,
    name: undefined as unknown as string,
  }

  event2.user = {
    id: message.elements[0]!.grayTipElement!.groupElement!.memberUin!,
    name: message.elements[0]!.grayTipElement!.groupElement!.memberNick!,
    avatar: `http://q.qlogo.cn/headimg_dl?dst_uin=${
      message.elements[0]!.grayTipElement!.groupElement!.memberUin
    }&spec=640`,
  }

  if (event2.member) delete event2.member

  return [event2, ...extraEvents]
}

/**
 * 解析他人被禁言消息。
 */
async function parseGuildMemberMuteMessage(
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  event: Event,
  message: RedMessage,
) {
  const [event2, extraEvents] = await parseChatMessage(
    self_id,
    config,
    event,
    message,
  )
  if (
    Number(message.elements[0]!.grayTipElement!.groupElement!.shutUp!.duration)
  )
    event2.type = 'chrono-unsafe-guild-mute'
  else event2.type = 'chrono-unsafe-guild-unmute'

  event2.operator = {
    id: message.elements[0]!.grayTipElement!.groupElement!.shutUp!.admin.uin,
    name: undefined as unknown as string,
  }

  event2.user = {
    id: message.elements[0]!.grayTipElement!.groupElement!.shutUp!.member.uin,
    name: message.elements[0]!.grayTipElement!.groupElement!.shutUp!.member
      .name,
    avatar: `http://q.qlogo.cn/headimg_dl?dst_uin=${
      message.elements[0]!.grayTipElement!.groupElement!.shutUp!.member.uin
    }&spec=640`,
  }

  if (event2.member) delete event2.member

  return [event2, ...extraEvents]
}

const regexGuildMemberAddedLegacyInviteMessage = /jp="(\d+)".*jp="(\d+)"/gim

/**
 * 解析旧版群成员邀请新人入群消息。使用 NT
 * 以前的客户端邀请他人加群会收到此消息。
 *
 * @remarks
 * 遗憾地，旧版群成员邀请新人入群消息不存在能够直接获取成员 QQ
 * 号的方法。需要通过解析 `grayTipElement`（elementType = 8）中的
 * `xmlElement`（subElementType = 12）中的 HTML 来提取 QQ 号。
 *
 * 一个 HTML 的示例：
 *
 * ```html
 * <gtip align="center">
 *   <qq uin="u_0gvBEjIEEOk5-EypJRjwxw" col="3" jp="1302744182" />
 *   <nor txt="邀请"/>
 *   <qq uin="u_ENIKFfFS74WSiKNoA6ERWg" col="3" jp="2953529126" />
 *   <nor txt="加入了群聊。"/>
 * </gtip>
 * ```
 *
 * 目前使用正则进行 QQ 号的提取。如果未来正则失效，则应考虑使用 cheerio
 * 进行提取。
 */
async function parseGuildMemberAddedLegacyInviteMessage(
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  event: Event,
  message: RedMessage,
) {
  const [event2, extraEvents] = await parseChatMessage(
    self_id,
    config,
    event,
    message,
  )
  event2.type = 'guild-member-added'

  const execArr = regexGuildMemberAddedLegacyInviteMessage.exec(
    message.elements[0]!.grayTipElement!.xmlElement!.content!,
  )
  if (!Array.isArray(execArr) || execArr.length < 3) return undefined
  const [_, operatorId, userId] = execArr

  event2.operator = {
    id: operatorId,
    name: undefined as unknown as string,
  }

  event2.user = {
    id: userId,
    name: undefined as unknown as string,
    avatar: `http://q.qlogo.cn/headimg_dl?dst_uin=${userId}&spec=640`,
  }

  if (event2.member) delete event2.member

  return [event2, ...extraEvents]
}

/**
 * 解析消息元素。
 */
async function parseElements(
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  message: RedMessage,
) {
  const elements: Element[] = []
  const extraEvents: Event[] = []

  for (const m of message.elements) {
    switch (m.elementType) {
      case 1: {
        // 文本消息
        switch (m.textElement!.atType) {
          case 0: {
            // 纯文本消息
            elements.push(Element.text(m.textElement?.content))
            break
          }

          case 2: {
            // at 消息
            const id = m.textElement!.atNtUin
            const name = m.textElement!.content.slice(1)
            if (!id) {
              extraEvents.push({
                id: undefined as unknown as number,
                platform: config.platform!,
                self_id,
                timestamp: Number(message.msgTime) * 1000,

                type: 'chrono-unsafe-warning-2131',
                user: {
                  id: undefined as unknown as string,
                  name,
                },
              })
              break
            }
            elements.push(
              Element('at', {
                id,
                name,
              }),
            )
            break
          }
        }
        break
      }

      case 2: {
        // 图片消息
        elements.push(
          Element('img', {
            src: `${config.self_url}/v1/assets/${Buffer.from(
              JSON.stringify({
                msgId: message.msgId,
                chatType: message.chatType,
                peerUid: message.peerUid,
                elementId: m.elementId,
                thumbSize: m.picElement!.thumbFileSize,
              }),
            ).toString('base64url')}`,
            'chrono-unsafe-isemoji':
              m.picElement!.picSubType === 1 || undefined,
            'chrono-unsafe-filepath': m.picElement!.sourcePath,
            'chrono-unsafe-filetype': m.picElement!.picType,
          }),
        )
        break
      }

      case 3: {
        // 文件消息
        elements.push(
          Element('file', {
            src: `${config.self_url}/v1/assets/${Buffer.from(
              JSON.stringify({
                msgId: message.msgId,
                chatType: message.chatType,
                peerUid: message.peerUid,
                elementId: m.elementId,
                thumbSize: m.fileElement!.thumbFileSize,
              }),
            ).toString('base64url')}`,
            'chrono-unsafe-filename': m.fileElement!.fileName,
          }),
        )
        break
      }

      case 4: {
        // 语音消息
        elements.push(
          Element('audio', {
            src: `${config.self_url}/v1/assets/${Buffer.from(
              JSON.stringify({
                msgId: message.msgId,
                chatType: message.chatType,
                peerUid: message.peerUid,
                elementId: m.elementId,
                thumbSize: 0,
              }),
            ).toString('base64url')}`,
            'chrono-unsafe-filepath': m.pttElement!.filePath,
            'chrono-unsafe-filename': m.pttElement!.fileName,
            'chrono-unsafe-time': m.pttElement!.duration,
            'chrono-unsafe-voice-change-type': m.pttElement!.voiceChangeType,
            'chrono-unsafe-wave-amplitudes':
              m.pttElement!.waveAmplitudes.join(),
          }),
        )
        break
      }

      case 5: {
        // 视频消息
        elements.push(
          Element('video', {
            src: `${config.self_url}/v1/assets/${Buffer.from(
              JSON.stringify({
                msgId: message.msgId,
                chatType: message.chatType,
                peerUid: message.peerUid,
                elementId: m.elementId,
                thumbSize: m.videoElement!.thumbSize,
              }),
            ).toString('base64url')}`,
            'chrono-unsafe-filepath': m.videoElement!.filePath,
            'chrono-unsafe-filename': m.videoElement!.fileName,
            'chrono-unsafe-file-format': m.videoElement!.fileFormat,
            'chrono-unsafe-time': m.videoElement!.fileTime,
          }),
        )
        break
      }

      default:
        break
    }
  }

  return [elements, extraEvents] as const
}
