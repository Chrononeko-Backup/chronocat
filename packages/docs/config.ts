import type { Options, ThemeConfig } from '@docusaurus/preset-classic'
import type { Config } from '@docusaurus/types'
import { join } from 'node:path'
import type { PresetOptions } from 'redocusaurus'

const title = 'Chronocat'
const tagline = '小巧轻便的 Red 实现'

const logo = {
  alt: 'Chronocat Logo',
  src: 'chronocat.svg',
  href: '/',
}

const presetConfig: Options = {
  docs: {
    routeBasePath: '/',
    sidebarPath: join(__dirname, 'sidebars.ts'),
    editUrl:
      'https://github.com/BuDingOwO/chronocat/tree/master/packages/docs',
  },

  blog: {
    postsPerPage: 'ALL',
    blogSidebarCount: 0,
  },

  theme: {
    customCss: join(__dirname, 'src/css/custom.scss'),
  },
}

const redocusaurusConfig: PresetOptions = {
  specs: [
    {
      id: 'openapi',
      spec: 'static/openapi.yaml',
      url: 'https://chronocat-docs.vercel.app/openapi.yaml',
      route: '/develop/satori/reference/api',
      layout: {
        title: 'API 参考',
      },
    },
    {
      id: 'config',
      spec: 'static/openapi-config.json',
      url: 'https://chronocat-docs.vercel.app/openapi-config.json',
    },
  ],
  theme: {
    primaryColor: '#2e8555',
    primaryColorDark: '#25c2a0',
  },
}

const themeConfig: ThemeConfig = {
  colorMode: {
    defaultMode: 'light',
    respectPrefersColorScheme: true,
  },

  image: 'chronocat.svg',

  metadata: [
    {
      name: 'description',
      content: tagline,
    },
    {
      name: 'keywords',
      content:
        'Chronocat, Red, Chrononeko, RedProtocol, BetterQQNT, LiteLoaderQQNT, QQNTim, NoneBot, NoneBot2, Koishi, Koishi.js, Bot, Chatbot, 机器人, OneBot, QQ',
    },
  ],

  navbar: {
    title,

    logo,

    items: [
      {
        label: '使用',
        position: 'left',
        type: 'dropdown',
        to: 'install',
        items: [
          {
            label: '安装',
            to: 'install',
          },
          {
            label: '启动',
            to: 'launch',
          },
          {
            label: '接入',
            to: 'connect',
          },
          {
            label: '配置',
            to: 'config',
          },
        ],
      },
      {
        label: '开发',
        position: 'left',
        to: 'develop',
      },
      {
        label: '贡献',
        position: 'left',
        to: 'contribute',
      },
      {
        label: '更多',
        position: 'left',
        type: 'dropdown',
        items: [
          {
            label: 'FAQ',
            to: 'faq/1',
          },
          {
            label: '博客',
            to: 'blog',
          },
          {
            label: '更多内容',
            to: 'more',
          },
        ],
      },
      {
        href: 'https://github.com/BuDingOwO/chronocat',
        html: '<svg style="width:24px;height:24px;display:block" viewBox="0 0 24 24"><path fill="currentColor" d="M12,2A10,10 0 0,0 2,12C2,16.42 4.87,20.17 8.84,21.5C9.34,21.58 9.5,21.27 9.5,21C9.5,20.77 9.5,20.14 9.5,19.31C6.73,19.91 6.14,17.97 6.14,17.97C5.68,16.81 5.03,16.5 5.03,16.5C4.12,15.88 5.1,15.9 5.1,15.9C6.1,15.97 6.63,16.93 6.63,16.93C7.5,18.45 8.97,18 9.54,17.76C9.63,17.11 9.89,16.67 10.17,16.42C7.95,16.17 5.62,15.31 5.62,11.5C5.62,10.39 6,9.5 6.65,8.79C6.55,8.54 6.2,7.5 6.75,6.15C6.75,6.15 7.59,5.88 9.5,7.17C10.29,6.95 11.15,6.84 12,6.84C12.85,6.84 13.71,6.95 14.5,7.17C16.41,5.88 17.25,6.15 17.25,6.15C17.8,7.5 17.45,8.54 17.35,8.79C18,9.5 18.38,10.39 18.38,11.5C18.38,15.32 16.04,16.16 13.81,16.41C14.17,16.72 14.5,17.33 14.5,18.26C14.5,19.6 14.5,20.68 14.5,21C14.5,21.27 14.66,21.59 15.17,21.5C19.14,20.16 22,16.42 22,12A10,10 0 0,0 12,2Z" /></svg>',
        position: 'right',
      },
    ],
  },

  footer: {
    style: 'dark',

    logo: {
      ...logo,
      width: 80,
    },

    links: [
      {
        title: '使用',
        items: [
          {
            label: '安装',
            to: '/install',
          },
          {
            label: '启动',
            to: '/launch',
          },
          {
            label: '接入',
            to: '/connect',
          },
          {
            label: '配置',
            to: '/config',
          },
        ],
      },
      {
        title: '开发与贡献',
        items: [
          {
            label: 'Satori',
            to: '/develop/satori',
          },
          {
            label: 'RedProtocol',
            to: '/develop/red',
          },
          {
            label: '贡献',
            to: '/contribute',
          },
        ],
      },
      {
        title: '更多',
        items: [
          {
            label: 'FAQ',
            to: '/faq/1',
          },
          {
            label: '博客',
            to: '/blog',
          },
          {
            label: '更多内容',
            to: '/more',
          },
          {
            label: '全站搜索',
            to: '/search',
          },
        ],
      },
    ],
    copyright: `<span>Copyright © ${new Date().getFullYear()} Team Chrononeko</span>`,
  },

  docs: {
    sidebar: {
      autoCollapseCategories: true,
    },
  },

  prism: {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    theme: require('prism-react-renderer/themes/github'),
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    darkTheme: require('prism-react-renderer/themes/palenight'),
  },
}

export const config: Config = {
  title,
  tagline,

  url: 'https://chronocat-docs.vercel.app',
  baseUrl: '/',

  onBrokenLinks: 'throw',
  onBrokenMarkdownLinks: 'warn',

  favicon: 'chronocat.ico',

  organizationName: 'chrononeko',
  projectName: 'chronocat',

  i18n: {
    defaultLocale: 'zh-Hans',
    locales: ['zh-Hans'],
  },

  presets: [
    ['classic', presetConfig],
    ['redocusaurus', redocusaurusConfig],
  ],

  plugins: ['docusaurus-plugin-sass'],

  themes: [
    [
      require.resolve('@easyops-cn/docusaurus-search-local'),
      {
        hashed: true,
        language: ['en', 'zh'],
        docsRouteBasePath: '/',
      },
    ],
  ],

  themeConfig,
}
