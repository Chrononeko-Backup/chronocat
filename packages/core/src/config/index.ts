import { load } from 'js-yaml'
import { readFile } from 'node:fs/promises'
import { join } from 'node:path'
import { getAuthData } from '../utils/authData'
import { baseDir } from '../utils/baseDir'
import { ensureConfig } from './ensure'
import { parseConfig } from './parser'
import type { ChronocatCurrentConfig } from './types'

let config: ChronocatCurrentConfig | undefined = undefined

const loadConfig = async () => {
  const configDir = join(baseDir, 'config')
  const configPath = join(configDir, 'chronocat.yml')

  await ensureConfig()

  const config = load(await readFile(configPath, 'utf-8'))
  const { uin } = await getAuthData()

  return parseConfig(config, uin)
}

export const getConfig = async () => {
  if (!config) config = await loadConfig()
  return config
}
