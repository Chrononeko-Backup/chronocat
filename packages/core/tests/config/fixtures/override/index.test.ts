import { load } from 'js-yaml'
import { readFile } from 'node:fs/promises'
import { join } from 'node:path'
import { parseConfig } from '../../../../src/config/parser'

test('配置文件解析器应当正确解析 被覆写的默认值', async () => {
  expect.assertions(2)

  const raw = (await readFile(join(__dirname, 'chronocat.yml'))).toString()
  const config = load(raw)

  const parsed1 = parseConfig(config, '1')
  expect(parsed1).toMatchSnapshot()

  const parsed10000 = parseConfig(config, '10000')
  // 虽然 enable 的默认值为 true，但由于 10000 账号的配置并未覆写 enable，所以解析得到的 enable 仍应为 false
  expect(parsed10000).toMatchSnapshot()
})
