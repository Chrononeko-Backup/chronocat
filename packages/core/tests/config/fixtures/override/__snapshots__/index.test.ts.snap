// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`配置文件解析器应当正确解析 被覆写的默认值 1`] = `
{
  "enable": false,
  "servers": [
    {
      "enable": true,
      "listen": "0.0.0.0",
      "port": 16530,
      "token": "DEFINE_CHRONO_TOKEN",
      "type": "red",
    },
    {
      "enable": true,
      "listen": "0.0.0.0",
      "platform": "chronocat",
      "port": 5500,
      "self_url": "https://chronocat.vercel.app",
      "token": "DEFINE_CHRONO_TOKEN",
      "type": "satori",
    },
  ],
}
`;

exports[`配置文件解析器应当正确解析 被覆写的默认值 2`] = `
{
  "enable": false,
  "servers": [
    {
      "enable": true,
      "listen": "0.0.0.0",
      "port": 16531,
      "token": "DEFINE_CHRONO_TOKEN",
      "type": "red",
    },
    {
      "enable": true,
      "listen": "0.0.0.0",
      "platform": "chronocat",
      "port": 5501,
      "self_url": "https://chronocat.vercel.app",
      "token": "DEFINE_CHRONO_TOKEN",
      "type": "satori",
    },
  ],
}
`;
